src/bidiMapper/domains/context/browsingContextProcessor.ts

```diff
@@ -286,46 +294,77 @@ export class BrowsingContextProcessor {
-    const scripts: Script.PreloadScript[] = [];
+    const cdpTargets = new Set<CdpTarget>(
+      params.context === undefined || params.context === null
+        ? this.#browsingContextStorage
+            .getTopLevelContexts()
+            .map((context) => context.cdpTarget)
+        : [this.#browsingContextStorage.getContext(params.context).cdpTarget]
+    );
 
-    if (params.context) {
-      // TODO(#293): Handle edge case with OOPiF. Whenever a frame is moved out
-      // of process, we have to add those scripts as well.
-      contexts.push(this.#browsingContextStorage.getContext(params.context));
-    } else {
-      // Add all contexts.
-      // TODO(#293): Add preload scripts to all new browsing contexts as well.
-      contexts.push(...this.#browsingContextStorage.getAllContexts());
+    const cdpPreloadScripts: CdpPreloadScript[] = [];
+
+    for (const cdpTarget of cdpTargets) {
+      const cdpPreloadScriptId = await cdpTarget.addPreloadScript(
+        // The spec provides a function, and CDP expects an evaluation.
+        `(${params.expression})();`,
+        params.sandbox
+      );
+      cdpPreloadScripts.push({
+        target: cdpTarget,
+        preloadScriptId: cdpPreloadScriptId,
+      });
     }
 
-    scripts.push(
-      ...(await Promise.all(
-        contexts.map((context) =>
-          context.cdpTarget.addPreloadScript(
-            // The spec provides a function, and CDP expects an evaluation.
-            `(${params.expression})();`,
-            params.sandbox
-          )
-        )
-      ))
-    );
+    const preloadScript: BidiPreloadScript =
+      this.#preloadScriptStorage.addPreloadScripts(
+        params.context ?? null,
+        cdpPreloadScripts,
+        params.expression
+      );
 
-    // TODO(#293): What to return whenever there are multiple contexts?
     return {
       result: {
-        script: scripts[0]!,
+        script: preloadScript.id,
       },
     };
   }
 
-  // eslint-disable-next-line @typescript-eslint/require-await
   async process_script_removePreloadScript(
-    _params: Script.RemovePreloadScriptParameters
+    params: Script.RemovePreloadScriptParameters
   ): Promise<Script.RemovePreloadScriptResult> {
-    throw new Message.UnknownErrorException('Not implemented.');
+    const bidiId = params.script;
+
+    const scripts = this.#preloadScriptStorage.findPreloadScripts({
+      id: bidiId,
+    });
 
-    return {};
+    if (scripts.length === 0) {
+      throw new Message.NoSuchScriptException(
+        `No preload script with BiDi ID '${bidiId}'`
+      );
+    }
+
+    for (const script of scripts) {
+      const contextId = script.contextId;
+
+      for (const cdpPreloadScript of script.cdpPreloadScripts) {
+        const cdpTarget = cdpPreloadScript.target;
+        const cdpPreloadScriptId = cdpPreloadScript.preloadScriptId;
+
+        if (contextId !== null) {
+          // Ensure context exists. TODO: Is this necessary?
```

Comment:
No need in checking if the `contextId` exists. Just verify the `cdpTarget` is there (IDK how exactly. It will probably throw an exception if it's session is closed).

Reply:
Done